libs/utils/src/properties.c

```diff
@@ -103,12 +132,275 @@ static void updateBuffers(char **key, char ** value, char **output, int outputPo
+        // entry is part of the properties entries buffer -> nop.
+    } else {
+        free(entry);
+    }
+}
+
+/**
+ * Create and add entry and optionally use the short properties optimization buffers.
+ * Only 1 of the types values (strValue, LongValue, etc) should be provided.
+ */
+static celix_status_t celix_properties_createAndSetEntry(celix_properties_t* properties,
+                                                         const char* key,
+                                                         const char** strValue,
+                                                         const long* longValue,
+                                                         const double* doubleValue,
+                                                         const bool* boolValue,
+                                                         celix_version_t* versionValue) {
+    if (!properties) {
+        return CELIX_SUCCESS; // silently ignore
+    }
+    if (!key) {
+        celix_err_push("Cannot set property with NULL key");
+        return CELIX_SUCCESS; // print error and ignore
+    }
+
+    celix_properties_entry_t* entry =
+        celix_properties_createEntry(properties, strValue, longValue, doubleValue, boolValue, versionValue);
+    if (!entry) {
+        return CELIX_ENOMEM;
+    }
+
+    const char* mapKey = key;
+    if (!celix_stringHashMap_hasKey(properties->map, key)) {
+        // new entry, needs new allocated key;
+        mapKey = celix_properties_createString(properties, key);
+        if (!mapKey) {
+            celix_properties_destroyEntry(properties, entry);
+            return CELIX_ENOMEM;
+        }
+    }
+
+    celix_status_t status = celix_stringHashMap_put(properties->map, mapKey, entry);
+    if (status != CELIX_SUCCESS) {
+        celix_properties_destroyEntry(properties, entry);
+        if (mapKey != key) {
+            celix_properties_freeString(properties, (char*)mapKey);
+        }
+    }
+    return status;
+}
+
+static void celix_properties_removeKeyCallback(void* handle, char* key) {
+    celix_properties_t* properties = handle;
+    celix_properties_freeString(properties, key);
+}
+
+static void celix_properties_removeEntryCallback(void* handle,
+                                                 const char* key __attribute__((unused)),
+                                                 celix_hash_map_value_t val) {
+    celix_properties_t* properties = handle;
+    celix_properties_entry_t* entry = val.ptrValue;
+    celix_properties_destroyEntry(properties, entry);
+}
+
+celix_properties_t* celix_properties_create() {
+    celix_properties_t* props = malloc(sizeof(*props));
+    if (props != NULL) {
+        celix_string_hash_map_create_options_t opts = CELIX_EMPTY_STRING_HASH_MAP_CREATE_OPTIONS;
+        opts.storeKeysWeakly = true;
+        opts.initialCapacity = (unsigned int)ceil(CELIX_SHORT_PROPERTIES_OPTIMIZATION_ENTRIES_SIZE / 0.75);
```

Comment:
As remarked previously, 16 slots is enough for 80-160 entries for separate chaining hash table.

Reply:
updated to 8.